tools/wake/timeline.cpp

```diff
@@ -0,0 +1,242 @@
+/* Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <algorithm>
+#include <map>
+#include <fstream>
+
+#include "timeline.h"
+#include "runtime/database.h"
+#include "util/execpath.h"
+
+struct JobNode {
+    JobReflection &job;
+    long parent_id = -1;
+    std::vector<long> dependencies;
+
+    long long starttime;
+    long long endtime;
+
+    long long hypothetical_starttime = 0;
+    long long hypothetical_endtime = -1;
+
+    bool top_sort_visited = false;
+    bool is_on_critical_path = false;
+    bool transitive_reduction_visited = false;
+
+    explicit JobNode(JobReflection &_job) : job(_job) {
+        std::string start_str = std::to_string(_job.starttime.asInt64());
+        std::string end_str = std::to_string(_job.endtime.asInt64());
+        starttime = std::stoll(start_str.substr(0, start_str.length() - 6));
+        endtime = std::stoll(end_str.substr(0, end_str.length() - 6));
+    }
+};
+
+void dfs_top_sort(long job_id, std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    JobNode &job = job_map.at(job_id);
+    job_map.at(job_id).top_sort_visited = true;
+    for (auto it = job.dependencies.begin(); it < job.dependencies.end(); it++) {
+        if (job_map.find(*it) == job_map.end()) {
+            job.dependencies.erase(it--);
+            continue;
+        }
+        if (!(job_map.at(*it).top_sort_visited)) {
+            dfs_top_sort(*it, job_map, top_sorted_jobs);
+        }
+    }
+    top_sorted_jobs.emplace_back(job_id);
+}
+
+void top_sort(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
```

Comment:
You can return the vector here I think? Unless there's a reason you'd want to call this with an already populated vector.

Reply:
Done